src/osd/PrimaryLogPG.cc

```diff
@@ -2849,6 +2860,47 @@ PrimaryLogPG::cache_result_t PrimaryLogPG::maybe_handle_cache_detail(
     do_proxy_read(op);
     return cache_result_t::HANDLED_PROXY;
 
+  case pg_pool_t::CACHEMODE_SWAP:
+  {
+    if (must_promote ||
+        (agent_state->promote_mode == TierAgentState::PROMOTE_MODE_WARMING &&
+	 !op->need_skip_promote())) {
+      if (agent_state &&
+          agent_state->evict_mode == TierAgentState::EVICT_MODE_FULL) {
+        dout(20) << __func__ << " cache pool full, waiting" << dendl;
+        block_write_on_full_cache(missing_oid, op);
+        return cache_result_t::BLOCKED_FULL;
+      }
+      agent_state->promote_queue.erase(missing_oid);
+      promote_object(obc, missing_oid, oloc, op, promote_obc);
+      return cache_result_t::BLOCKED_PROMOTE;
+    }
+    if (op->may_write() || op->may_cache() || write_ordered) {
+      do_proxy_write(op);
+    } else {
+      do_proxy_read(op);
+      if (obc.get() && obc->is_blocked()) {
+	if (promote_obc)
+	  *promote_obc = obc;
+	return cache_result_t::BLOCKED_PROMOTE;
+      }
+    }
+    auto p = agent_state->promote_queue.find(missing_oid);
+    if (p == agent_state->promote_queue.end())
+      return cache_result_t::HANDLED_PROXY;
+    else if (p->second.temp < agent_state->promote_temp) {
+      agent_state->promote_queue.erase(p);
+      return cache_result_t::HANDLED_PROXY;
+    } else if (!op->need_skip_promote() &&
+	agent_state->promote_mode == TierAgentState::PROMOTE_MODE_SOME &&
+        !osd->promote_throttle()) {
+      agent_state->promote_queue.erase(missing_oid);
+      promote_object(obc, missing_oid, oloc, op, promote_obc);
+      return cache_result_t::BLOCKED_PROMOTE;
+    }
+    return cache_result_t::HANDLED_PROXY;
+  }
```

Comment:
I don't really understand this strategy.  It looks like:

#. on reply, we note temperature, and if it's high, we add to the promote_queue
#. on the next request for the object, we trigger a blocking promote

Is that right?  It seems like (1) it would be better to trigger the promote as soon as you see the high temperature so that it doesn't block the next write (or blocks less of it).  Also, (2) making promote_queue an actual linear queue means it can be ordered by time and the trimming will be efficient, and you won't need to lookup by hobject_t here.

Reply:
Hi Sage, thanks for your review and suggestion.

I've tried three strategy:
1. Promote objects immediately in finisher thread
2. Promote objects in agent thread periodly
3. Promote object on the next request for the object

The first strategy is much more reasonable. The only concern is that the efficiency of finisher thread may be reduced. So I choose the third one to keep the similar logic as other cache modes.

I'll retest the first strategy and make changes.

Reply:
I think (3) is fine for now; I'm also a bit nervous about adding new behavior that varies to much from what we have now.

I was confused by the temperature being returned in MOSDOpReply... what is that used for?